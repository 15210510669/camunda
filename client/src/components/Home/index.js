/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

export {default as Home} from './Home';
export {default as Collection} from './Collection';
