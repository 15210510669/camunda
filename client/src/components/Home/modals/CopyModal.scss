.CopyModal {
  .moveSection {
    border-top: 1px solid var(--silver-darken-80);
  }
}
