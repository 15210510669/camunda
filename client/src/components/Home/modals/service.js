/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {get} from 'request';
import {t} from 'translation';

export async function searchIdentities(terms) {
  const response = await get(`api/identity/search`, {terms});

  return await response.json();
}

export async function getDefinitionTenants(defintionKey, defintionType) {
  const response = await get(`api/definition/${defintionType}/${defintionKey}`);

  return await response.json();
}

export async function getDefinitionsWithTenants() {
  const response = await get('api/definition');

  return await response.json();
}

export async function getTenantsWithDefinitions() {
  const response = await get('api/definition/_groupByTenant');

  return await response.json();
}

export function formatDefintionName({name, type}) {
  const typeLabel = type === 'process' ? t('home.sources.process') : t('home.sources.decision');

  return name + ' (' + typeLabel + ')';
}
