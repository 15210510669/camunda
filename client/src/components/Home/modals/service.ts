/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get} from 'request';
import {Definition, Tenant} from 'types';

export type TenantWithDefinitions = Tenant & {definitions: Definition};

export type DefinitionWithTenants = Omit<Definition, 'key'> & {tenants: Tenant[]; key: string};

export async function getDefinitionsWithTenants(): Promise<DefinitionWithTenants[]> {
  const response = await get('api/definition');

  return await response.json();
}

export async function getTenantsWithDefinitions(): Promise<TenantWithDefinitions[]> {
  const response = await get('api/definition/_groupByTenant');

  return await response.json();
}
