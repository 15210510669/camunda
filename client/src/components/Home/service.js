/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {get, post, put, del} from 'request';

export async function loadEntities() {
  const response = await get('api/entities');
  return await response.json();
}

export async function addUser(collection, id, type, role) {
  return await post(`api/collection/${collection}/role`, {identity: {id, type}, role});
}

export async function editUser(collection, id, role) {
  return await put(`api/collection/${collection}/role/${id}`, {role});
}

export async function removeUser(collection, id) {
  return await del(`api/collection/${collection}/role/${id}`);
}

export async function loadAlerts(collection) {
  const response = await get(`api/collection/${collection}/alerts`);
  return await response.json();
}

export async function addAlert(alert) {
  return await post(`api/alert`, alert);
}

export async function editAlert(id, alert) {
  return await put(`api/alert/${id}`, alert);
}

export async function removeAlert(id) {
  return await del(`api/alert/${id}`);
}

export async function copyEntity(type, id, name, collectionId) {
  const query = {name};

  if (collectionId || collectionId === null) {
    query.collectionId = collectionId;
  }

  const response = await post(`api/${type}/${id}/copy`, undefined, {query});
  const json = await response.json();

  return json.id;
}

export async function searchIdentities(terms) {
  const response = await get(`api/identity/search`, {terms});

  return await response.json();
}
