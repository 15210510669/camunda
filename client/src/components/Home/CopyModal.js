/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {Component} from 'react';

import {Button, LabeledInput, Modal, Form, Switch, Typeahead, InfoMessage} from 'components';
import {t} from 'translation';
import {withErrorHandling} from 'HOC';
import {showError} from 'notifications';

import {loadEntities} from './service';

import './CopyModal.scss';

export default withErrorHandling(
  class CopyModal extends Component {
    constructor(props) {
      super(props);

      this.state = {
        name: props.entity.name + ` (${t('common.copyLabel')})`,
        moving: false,
        availableCollections: [],
        collection: null,
        gotoNew: true
      };
    }

    componentDidMount() {
      if (this.props.entity.entityType !== 'collection') {
        this.props.mightFail(
          loadEntities(),
          entities =>
            this.setState({
              availableCollections: [
                {id: null, entityType: 'collection', name: t('navigation.homepage')},
                ...entities
              ].filter(
                ({entityType, id}) => entityType === 'collection' && id !== this.props.collection
              )
            }),
          showError
        );
      }
    }

    onConfirm = () => {
      const {name, moving, collection, gotoNew} = this.state;
      if (name && (!moving || collection)) {
        if (this.isCollection() && this.props.jumpToEntity) {
          this.props.onConfirm(name, gotoNew);
        } else {
          this.props.onConfirm(name, moving && gotoNew, moving && collection.id);
        }
      }
    };

    getMulticopyText = () => {
      const {entityType, data} = this.props.entity;
      const containedReports = data.subEntityCounts.report;

      if (containedReports) {
        return t(containedReports > 1 ? 'home.copy.subEntities' : 'home.copy.subEntity', {
          entityType: entityType === 'dashboard' ? t('dashboard.label') : t('home.types.combined'),
          number: containedReports
        });
      }
    };

    renderMoveOption = () => {
      const {moving, availableCollections, collection} = this.state;
      const multicopyText = this.getMulticopyText();

      return (
        <>
          <Form.Group className="moveSection">
            <Switch
              label={t('home.copy.moveLabel')}
              checked={moving}
              onChange={({target: {checked}}) => this.setState({moving: checked})}
            />
          </Form.Group>
          {moving && (
            <>
              <Form.Group noSpacing>
                <Typeahead
                  initialValue={collection}
                  noValuesMessage={t('home.copy.noCollections')}
                  placeholder={t('home.copy.pleaseSelect')}
                  values={availableCollections}
                  formatter={({name}) => name}
                  onSelect={collection => this.setState({collection})}
                />
              </Form.Group>
              {multicopyText && <Form.Group>{multicopyText}</Form.Group>}
            </>
          )}
        </>
      );
    };

    isCollection = () => this.props.entity.entityType === 'collection';

    render() {
      const {onClose, entity, jumpToEntity} = this.props;
      const {name, moving, collection, gotoNew} = this.state;

      return (
        <Modal className="CopyModal" open onClose={onClose} onConfirm={this.onConfirm}>
          <Modal.Header>{t('common.copyName', {name: entity.name})}</Modal.Header>
          <Modal.Content>
            <Form>
              <Form.Group>
                <LabeledInput
                  type="text"
                  label={t('home.copy.inputLabel')}
                  value={name}
                  autoComplete="off"
                  onChange={({target: {value}}) => this.setState({name: value})}
                />
              </Form.Group>
              {this.isCollection() && (
                <InfoMessage>{t('home.copy.copyCollectionInfo')}</InfoMessage>
              )}
              {!this.isCollection() && this.renderMoveOption()}
              {jumpToEntity && (this.isCollection() || moving) && (
                <Form.Group>
                  <LabeledInput
                    label={t('home.copy.gotoNew')}
                    type="checkbox"
                    checked={gotoNew}
                    onChange={({target: {checked}}) => this.setState({gotoNew: checked})}
                  />
                </Form.Group>
              )}
            </Form>
          </Modal.Content>
          <Modal.Actions>
            <Button className="cancel" onClick={onClose}>
              {t('common.cancel')}
            </Button>
            <Button
              disabled={!name || (moving && !collection)}
              variant="primary"
              color="blue"
              className="confirm"
              onClick={this.onConfirm}
            >
              {t('common.copy')}
            </Button>
          </Modal.Actions>
        </Modal>
      );
    }
  }
);
