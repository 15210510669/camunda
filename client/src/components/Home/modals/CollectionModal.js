/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useState} from 'react';
import {Redirect} from 'react-router-dom';
import {Button} from '@carbon/react';

import {LabeledInput, CarbonModal as Modal, Form} from 'components';
import {t} from 'translation';
import {withErrorHandling} from 'HOC';
import {showError} from 'notifications';

import {addSources} from '../service';
import SourcesModal from './SourcesModal';

import './CollectionModal.scss';

export function CollectionModal({
  onClose,
  title,
  confirmText,
  initialName,
  mightFail,
  onConfirm,
  showSourcesModal,
}) {
  const [name, setName] = useState(initialName);
  const [loading, setLoading] = useState(false);
  const [redirect, setRedirect] = useState(null);
  const [displaySourcesModal, setDisplaySourcesModal] = useState(false);

  const confirm = () => {
    if (!name || loading) {
      return;
    }

    if (showSourcesModal) {
      setDisplaySourcesModal(true);
    } else {
      performRequest();
    }
  };

  const performRequest = (sources) => {
    mightFail(
      onConfirm(name),
      (id) => {
        if (id && sources) {
          mightFail(
            addSources(id, sources),
            () => {
              setLoading(false);
              setRedirect(id);
            },
            showError
          );
        }
      },
      (error) => {
        showError(error);
        setLoading(false);
      }
    );

    setLoading(true);
  };

  if (redirect) {
    return <Redirect to={`/collection/${redirect}/`} />;
  }

  return (
    <>
      <Modal
        className="CollectionModal"
        open={!displaySourcesModal}
        onClose={onClose}
        onConfirm={confirm}
      >
        <Modal.Header>{title}</Modal.Header>
        <Modal.Content>
          <Form>
            {showSourcesModal && <div className="info">{t('common.collection.modal.info')}</div>}
            <Form.Group>
              <LabeledInput
                type="text"
                label={t('common.collection.modal.inputLabel')}
                style={{width: '100%'}}
                value={name}
                onChange={({target: {value}}) => setName(value)}
                disabled={loading}
                autoComplete="off"
                data-modal-primary-focus
              />
            </Form.Group>
          </Form>
        </Modal.Content>
        <Modal.Footer>
          <Button kind="secondary" className="cancel" onClick={onClose} disabled={loading}>
            {t('common.cancel')}
          </Button>
          <Button className="confirm" disabled={!name || loading} onClick={confirm}>
            {showSourcesModal ? t('common.collection.modal.addDataSources') : confirmText}
          </Button>
        </Modal.Footer>
      </Modal>
      {displaySourcesModal && (
        <SourcesModal
          onClose={onClose}
          onConfirm={performRequest}
          confirmText={confirmText}
          preSelectAll
        />
      )}
    </>
  );
}

export default withErrorHandling(CollectionModal);
