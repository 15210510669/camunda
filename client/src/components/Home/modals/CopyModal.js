/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';

import {Button, LabeledInput, Modal, Form, Message} from 'components';
import {t} from 'translation';
import MoveCopy from './MoveCopy';

import './CopyModal.scss';

export default class CopyModal extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.entity.name + ` (${t('common.copyLabel')})`,
      moving: false,
      collection: null,
      gotoNew: true
    };
  }

  onConfirm = () => {
    const {name, moving, collection, gotoNew} = this.state;
    if (name && (!moving || collection)) {
      if (this.isCollection() && this.props.jumpToEntity) {
        this.props.onConfirm(name, gotoNew);
      } else {
        this.props.onConfirm(name, moving && gotoNew, moving && collection.id);
      }
    }
  };

  isCollection = () => this.props.entity.entityType === 'collection';

  render() {
    const {onClose, entity, jumpToEntity} = this.props;
    const {name, moving, collection, gotoNew} = this.state;

    return (
      <Modal className="CopyModal" open onClose={onClose} onConfirm={this.onConfirm}>
        <Modal.Header>{t('common.copyName', {name: entity.name})}</Modal.Header>
        <Modal.Content>
          <Form>
            <Form.Group>
              <LabeledInput
                type="text"
                label={t('home.copy.inputLabel')}
                value={name}
                autoComplete="off"
                onChange={({target: {value}}) => this.setState({name: value})}
              />
            </Form.Group>
            {this.isCollection() && <Message>{t('home.copy.copyCollectionInfo')}</Message>}
            {!this.isCollection() && (
              <MoveCopy
                entity={entity}
                parentCollection={this.props.collection}
                moving={moving}
                setMoving={moving => this.setState({moving})}
                collection={collection}
                setCollection={collection => this.setState({collection})}
              />
            )}
            {jumpToEntity && (this.isCollection() || moving) && (
              <Form.Group>
                <LabeledInput
                  label={t('home.copy.gotoNew')}
                  type="checkbox"
                  checked={gotoNew}
                  onChange={({target: {checked}}) => this.setState({gotoNew: checked})}
                />
              </Form.Group>
            )}
          </Form>
        </Modal.Content>
        <Modal.Actions>
          <Button className="cancel" onClick={onClose}>
            {t('common.cancel')}
          </Button>
          <Button
            disabled={!name || (moving && !collection)}
            variant="primary"
            color="blue"
            className="confirm"
            onClick={this.onConfirm}
          >
            {t('common.copy')}
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }
}
