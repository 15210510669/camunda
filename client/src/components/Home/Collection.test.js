/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import {Dropdown, ConfirmationModal} from 'components';
import {refreshBreadcrumbs} from 'components/navigation';
import {loadEntity, deleteEntity, updateEntity} from 'services';

import CollectionWithErrorHandling from './Collection';
import CollectionModal from './CollectionModal';
import CopyModal from './CopyModal';
import {copyEntity} from './service';

jest.mock('./service', () => ({copyEntity: jest.fn()}));

const Collection = CollectionWithErrorHandling.WrappedComponent;

jest.mock('components/navigation', () => ({refreshBreadcrumbs: jest.fn()}));

jest.mock('services', () => {
  const rest = jest.requireActual('services');
  return {
    ...rest,
    deleteEntity: jest.fn(),
    updateEntity: jest.fn(),
    loadEntity: jest.fn().mockReturnValue({
      id: 'aCollectionId',
      name: 'aCollectionName',
      lastModified: '2017-11-11T11:11:11.1111+0200',
      created: '2017-11-11T11:11:11.1111+0200',
      owner: 'user_id',
      lastModifier: 'user_id',
      currentUserRole: 'manager', // or editor, viewer
      data: {
        configuration: {},
        entities: [
          {
            id: 'aDashboardId',
            name: 'aDashboard',
            lastModified: '2017-11-11T11:11:11.1111+0200',
            created: '2017-11-11T11:11:11.1111+0200',
            owner: 'user_id',
            lastModifier: 'user_id',
            entityType: 'dashboard',
            currentUserRole: 'editor', // or viewer
            data: {
              subEntityCounts: {
                report: 8
              }
            }
          },
          {
            id: 'aReportId',
            name: 'aReport',
            lastModified: '2017-11-11T11:11:11.1111+0200',
            created: '2017-11-11T11:11:11.1111+0200',
            owner: 'user_id',
            lastModifier: 'user_id',
            reportType: 'process', // or "decision"
            combined: false,
            entityType: 'report',
            currentUserRole: 'editor' // or viewer
          }
        ],
        roles: [
          {
            identity: {
              id: 'kermit',
              type: 'user' // or group
            },
            role: 'manager' // or editor, viewer
          }
        ], // array of role objects, for details see role endpoints
        scope: [] // array of scope objects, for details see scope endpoints
      }
    })
  };
});

const props = {
  mightFail: jest.fn().mockImplementation((data, cb) => cb(data)),
  match: {params: {id: 'aCollectionId'}}
};

it('should match snapshot', () => {
  const node = shallow(<Collection {...props} />);

  expect(node).toMatchSnapshot();
});

it('should show delete modal when clicking delete button', () => {
  const node = shallow(<Collection {...props} />);

  node
    .find(Dropdown.Option)
    .at(2)
    .simulate('click');

  expect(node.find(ConfirmationModal).prop('open')).toBe(true);
});

it('should delete collection', () => {
  const node = shallow(<Collection {...props} />);

  node.find(ConfirmationModal).prop('onConfirm')();

  expect(deleteEntity).toHaveBeenCalledWith('collection', 'aCollectionId');
  expect(node).toMatchSnapshot();
});

it('should show an edit modal when clicking the edit button', () => {
  const node = shallow(<Collection {...props} />);

  node
    .find(Dropdown.Option)
    .at(0)
    .simulate('click');

  expect(node.find(CollectionModal)).toExist();
});

it('should modify the collections name with the edit modal', async () => {
  const node = shallow(<Collection {...props} />);

  node.setState({editingCollection: true});
  await node.find(CollectionModal).prop('onConfirm')('new Name');

  expect(updateEntity).toHaveBeenCalledWith('collection', 'aCollectionId', {name: 'new Name'});
  expect(refreshBreadcrumbs).toHaveBeenCalled();
});

it('should hide edit/delete from context menu for collection items that does not have a "manager" role', () => {
  loadEntity.mockReturnValueOnce({
    id: 'aCollectionId',
    name: 'aCollectionName',
    owner: 'user_id',
    lastModifier: 'user_id',
    currentUserRole: 'editor',
    data: {
      entities: []
    }
  });
  const node = shallow(<Collection {...props} />);

  expect(node.find('.name').find({type: 'delete'})).not.toExist();
  expect(node.find('.name').find({type: 'edit'})).not.toExist();
});

it('should render content depending on the selected tab', () => {
  const node = shallow(
    <Collection {...props} match={{params: {id: 'aCollectionId', viewMode: 'users'}}} />
  );

  expect(node).toMatchSnapshot();
});

it('should show the copy modal when clicking the copy button', () => {
  const node = shallow(<Collection {...props} />);

  node
    .find(Dropdown.Option)
    .at(1)
    .simulate('click');

  expect(node.find(CopyModal)).toExist();
});

it('should copy entity and redirect to collection', () => {
  copyEntity.mockReturnValue('copyCollectionID');
  const node = shallow(<Collection {...props} />);

  node
    .find(Dropdown.Option)
    .at(1)
    .simulate('click');

  node.find(CopyModal).prop('onConfirm')('new Name', true);

  expect(copyEntity).toHaveBeenCalledWith('collection', 'aCollectionId', 'new Name');

  expect(node.find('Redirect').props().to).toBe('/collection/copyCollectionID/');

  expect(node.find(CopyModal)).not.toExist();
});
