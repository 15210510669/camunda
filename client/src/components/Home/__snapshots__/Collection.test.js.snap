// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render content depending on the selected tab 1`] = `
<div
  className="Collection"
>
  <div
    className="header"
  >
    <Icon
      size="24"
      type="collection"
    />
    <span
      className="text"
      title="aCollectionName"
    >
      aCollectionName
    </span>
    <Dropdown
      label={
        <Icon
          size="24px"
          type="context-menu"
        />
      }
    >
      <ForwardRef(DropdownOption)
        onClick={[Function]}
      >
        <Icon
          type="edit"
        />
        Edit
      </ForwardRef(DropdownOption)>
      <ForwardRef(DropdownOption)
        onClick={[Function]}
      >
        <Icon
          type="copy-document"
        />
        Copy
      </ForwardRef(DropdownOption)>
      <ForwardRef(DropdownOption)
        onClick={[Function]}
      >
        <Icon
          type="delete"
        />
        Delete
      </ForwardRef(DropdownOption)>
    </Dropdown>
    <ul
      className="navigation"
    >
      <li
        className=""
      >
        <Link
          to="."
        >
          Dashboards & Reports
        </Link>
      </li>
      <li
        className=""
      >
        <Link
          to="alerts"
        >
          Alerts
        </Link>
      </li>
      <li
        className="active"
      >
        <Link
          to="users"
        >
          Users
        </Link>
      </li>
      <li
        className=""
      >
        <Link
          to="sources"
        >
          Data Sources
        </Link>
      </li>
    </ul>
  </div>
  <div
    className="content"
  >
    <UserListErrorHandler
      collection="aCollectionId"
      onChange={[Function]}
      readOnly={false}
    />
  </div>
  <DeleterErrorHandler
    checkConflicts={[Function]}
    entity={false}
    onClose={[Function]}
    onDelete={[Function]}
    type={false}
  />
  <CopierErrorHandler
    collection="aCollectionId"
    entity={null}
    onCancel={[Function]}
    onCopy={[Function]}
  />
</div>
`;
